### PR TITLE
Add capabilities 

### PR DESCRIPTION
**Pull request by baylesj**
This patch adds specifications for getting device capabilities through the open screen protocol. Specifically, what types of actions and media does the receiver support?

NOTE: This patch is based on @pthatcherg's remote playback pull request, so make sure to exclude those commits when diffing!

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -843,4 +1380,227 @@ result = (
   unknown-error: 199
 )
 
+; type key 17
+remote-playback-availability-request = {
+  request
+  1: [* text] ; urls
+  2: microseconds ; watch-duration
+  3: int ; watch-id
```
Comment (rwkeane) [inline index.bs]: Seems be better to use `uint`instead of `int` for consistency with `connection-id`, `request-id`

File: index.bs [hunk 2]
```diff
@@ -843,4 +1380,227 @@ result = (
...
+}
+
+; type key 18
+remote-playback-availability-response = {
+  response
+  url-availabilities: [* url-availability]
```
Comment (rwkeane) [inline index.bs]: integer key to minimize payload

File: index.bs [hunk 3]
```diff
@@ -843,4 +1380,227 @@ result = (
...
+    loading: 2
+    no-source: 3
+  ) ; loading
+  ? 4: &(
+    nothing: 0
+    metdata: 1
+    current: 2
+    future: 3
+    enough: 4
+  ) ; loaded
```
Comment (rwkeane) [inline index.bs]: Even if `3` and `4`'s types aren't used elsewhere, I think pulling them out into enums defined outside this map would be more readable. Plus the generated C++ code would be more readable and easier to use

File: index.bs [hunk 4]
```diff
@@ -843,4 +1380,227 @@ result = (
...
+}
+
+added-text-track = {
+  1: &(
+    subtitles
+    captions
+    descriptions
+    chapters
+    metadata
+  ) ; kind
```
Comment (rwkeane) [inline index.bs]: same as above - define the enum as a type outside the map

File: index.bs [hunk 5]
```diff
@@ -843,4 +1380,227 @@ result = (
...
+    decode-error: 3
+    source-not-supported: 4
+  )
+  message: text
+]
+
+track-state = (
+  1: text ; id
+  2: text ; label
+  3: text ; language
```
Comment (rwkeane) [inline index.bs]: Is language shared with `video-track-state` too?

File: index.bs [hunk 6]
```diff
@@ -843,4 +1380,227 @@ result = (
...
+)
+
+audio-track-state = {
+  track-state
+  4: bool ; enabled
+}
+
+video-track-state = {
+  track-state
+  5 : bool ; selected
```
Comment (rwkeane) [inline index.bs]: looks like this should be `4` not `5`

File: index.bs [hunk 7]
```diff
@@ -843,4 +1380,227 @@ result = (
...
+}
+
+track-track-state = {
+  track-state
+  6 : mode ; text-track-mode
```
Comment (rwkeane) [inline index.bs]: looks like this should be `4` not `6`

File: index.bs [hunk 8]
```diff
@@ -843,4 +1380,227 @@ result = (
...
+}
+
+; type key 118
+remote-playback-termination-response = {
+  response
+  1: &result ; result
+}
+
+; type key 119
+remote-playback-termination-event = {
```
Comment (rwkeane) [inline index.bs]: Do we still need ` 0: request-id ; request-id` added here, even if isn't a request or response?

File: index.bs [hunk 9]
```diff
@@ -843,4 +1380,227 @@ result = (
...
+  ? 22: [* text-track-state] text-tracks
+}
+
+added-text-track = {
+  1: &(
+    subtitles
+    captions
+    descriptions
+    chapters
+    metadata
```
Comment (rwkeane) [inline index.bs]: Give integer values

File: index.bs (hunk 1 above)
Comment (btolsch) [reply index.bs]: This comment and your others apply to Peter's PR (#134).  Unfortunately, it seems like GitHub doesn't have support for dependent PRs (or maybe just not from separate authors?).  Looking at the individual commits is easier for me than looking at the total diff.

File: index.bs [hunk 10]
```diff
@@ -284,31 +284,55 @@ mechanism for metadata discovery.
 Transport and metadata discovery with QUIC {#transport}
 =======================================================
 
-If a listening agent wants to connect to an advertising agent, or to
-learn further metadata about it, it initiates a [[QUIC]] connection to
-the IP and port from the SRV record.  Prior to authentication, a message may be
-exchanged (such as further metadata), but such info should be treated as
-unverified (such as indicating to a user that a display name of an
+If a listening agent ("controller") wants to connect to an advertising agent
```
Comment (btolsch) [inline index.bs]: I'm sure Peter and/or Mark will chime in, but I think they have avoided controller/receiver terminology is because it seems Presentation-API-centric.  I'm neutral about this though.

File: index.bs [hunk 11]
```diff
@@ -284,31 +284,55 @@ mechanism for metadata discovery.
...
 
 : model-name (optional)
-:: If the agent is a hardware device, the model name of
+:: If the receiver is a hardware device, the model name of
     the device.  This is used mainly for debugging purposes, but may be
     displayed to the user of the requesting agent.
 
+: audio-only (optional)
+:: Generally speaking, receivers are assumed to support streaming both
+audio and video. If an receiver wishes to indicate that they only support audio
```
Comment (btolsch) [inline index.bs]: s/an receiver/a receiver/

aside: Oh man, did GitHub remove the suggestion feature?  That was really nice for typos like this.

File: index.bs [hunk 12]
```diff
@@ -284,31 +284,55 @@ mechanism for metadata discovery.
...
+following properties:
+
+: supported-resolutions (required)
+:: The receiver shall indicate what resolutions and refresh rates it supports
+streaming content at.
+
+<!-- TODO(jophba): figure out the units for this -->
+: max-pixel-fill-rate (optional)
+:: In combination with the supported resolutions, the receiver has the
+option to to reply with the supported pixel fill rate in <UNITS>.
```
Comment (btolsch) [inline index.bs]: s/to to/to/

File: index.bs [hunk 13]
```diff
@@ -1206,11 +1198,18 @@ resolution = {
   2: [1* float64] ; refresh-rates
 }
 
-<!-- TODO(jophba): what codecs do we support? -->
-<!-- TODO(jophba): do we need this? -->
-<!-- TODO(jophba): flag enum? -->
 codec = (
-
+  H-264-1080
```
Comment (btolsch) [inline index.bs]: These need to be `name: number`.

File: index.bs [hunk 14]
```diff
@@ -1206,11 +1198,18 @@ resolution = {
...
+  H-264-720
```
Comment (btolsch) [inline index.bs]: Why is there a separate 720 and 1080 h264?

File: index.bs (hunk 12 above)
Comment (baylesj) [reply index.bs]: Thx

File: index.bs (hunk 10 above)
Comment (baylesj) [reply index.bs]: I'll leave as is for now, and bring it up with them.

File: index.bs [hunk 15]
```diff
@@ -141,7 +141,7 @@ Presentation API Requirements {#requirements-presentation-api}
     {{PresentationConnection}} to an existing presentation on the
     receiver, given its [=presentation request URL=] and [=presentation ID=].
 
-6.  It must be possible to to close a {{PresentationConnection}} between a
```
Comment (pthatcherg) [inline index.bs]: About all these typos you found and fixed: could you make a separate PR that fixes those?  It would be nice to have them fixed independently of this PR.

File: index.bs (hunk 10 above)
Comment (pthatcherg) [reply index.bs]: I like controller/receiver as terms in general, but in the context of mDNS and discovery, we can't assume the receiver is always the advertiser and the controller is always the listener.

File: index.bs [hunk 16]
```diff
@@ -287,32 +287,85 @@ mechanism for metadata discovery.
...
+("receiver"), or to learn further metadata about it, it initiates a [[QUIC]]
+connection to the IP and port from the SRV record.  Prior to authentication,
+a message may be exchanged (such as further metadata), but such info should be
+treated as unverified (such as indicating to a user that a display name of an
 unauthenticated agent is unverified).
 
-To learn further metadata, an agent may send an agent-info-request
-message (see [[#appendix-a]]) and receive back an agent-info-response message.  The
-messages may contain the following information with the following meaning:
+To learn further metadata, the controller may send an agent-info-request
```
Comment (pthatcherg) [inline index.bs]: Actually, both sides will probably send this message to learn about the other side.

File: index.bs [hunk 17]
```diff
@@ -287,32 +287,85 @@ mechanism for metadata discovery.
...
-    responding agent changes its display name, the requesting agent should
-    make UI affordance for indicating to the user that the display name has
-    changed.
+:: The display name of the receiver, intended to be displayed to a user by the
+     controller. The controller should indicate through the UI if the receiver
+     is not authenticated or if the display name changes.
 
 : model-name (optional)
-:: If the agent is a hardware device, the model name of
+:: If the receiver is a hardware device, the model name of
```
Comment (pthatcherg) [inline index.bs]: The device might be a controller.  Imagine a TV wants to connect back to controller devices.

File: index.html [hunk 18]
```diff
@@ -1696,30 +1698,85 @@ <h2 class="heading settled" data-level="3" id="discovery"><span class="secno">3.
...
+a message may be exchanged (such as further metadata), but such info should be
+treated as unverified (such as indicating to a user that a display name of an
 unauthenticated agent is unverified).</p>
-   <p>To learn further metadata, an agent may send an agent-info-request
-message (see <a href="#appendix-a">Appendix A: Messages</a>) and receive back an agent-info-response message.  The
-messages may contain the following information with the following meaning:</p>
+   <p>To learn further metadata, the controller may send an agent-info-request
+message (see <a href="#appendix-a">Appendix A: Messages</a>) and receive back an agent-info-response message.
+The messages may contain the following properties:</p>
    <dl>
```
Comment (pthatcherg) [inline index.html]: Same here: we can't assume the agent is a controller.

File: index.html [hunk 19]
```diff
@@ -1540,7 +1540,9 @@ <h3 class="heading settled" data-level="1.1" id="terminology"><span class="secno
...
+a browser). <a data-link-type="dfn" href="https://w3c.github.io/presentation-api/#dfn-presentation-display-availability" id="ref-for-dfn-presentation-display-availability">presentation display availability</a> refers to whether or not a <a data-link-type="dfn" href="https://w3c.github.io/presentation-api/#dfn-receiver" id="ref-for-dfn-receiver①">receiver</a> is compatible with a <a data-link-type="dfn" href="https://w3c.github.io/presentation-api/#dfn-presentation-request-url" id="ref-for-dfn-presentation-request-url">presentation request URL</a>.  However, in the <a data-link-type="biblio" href="#biblio-presentation-api">Presentation API</a>, a "controller" refers to as a specific
+browsing context within the browser, whereas here the "controller" refers to the
+browser itself, although it may be acting on behalf of a browsing context.</p>
    <p>We borrow terminology from the <a data-link-type="biblio" href="#biblio-remote-playback">Remote Playback API</a>.  The
```
Comment (pthatcherg) [inline index.html]: Hmm... do you need to rebase again?  I'm seeing in your diff things I made in the last patch of my PR.

File: index.html [hunk 20]
```diff
@@ -1696,30 +1698,85 @@ <h2 class="heading settled" data-level="3" id="discovery"><span class="secno">3.
...
-responding agent changes its display name, the requesting agent should
-make UI affordance for indicating to the user that the display name has
-changed.</p>
+     <p>The display name of the receiver, intended to be displayed to a user by the
+ controller. The controller should indicate through the UI if the receiver
+ is not authenticated or if the display name changes.</p>
     <dt data-md>model-name (optional)
     <dd data-md>
-     <p>If the agent is a hardware device, the model name of
+     <p>If the receiver is a hardware device, the model name of
```
Comment (pthatcherg) [inline index.html]: Here again with assuming receiver == hardware

File: index.html [hunk 21]
```diff
@@ -1696,30 +1698,85 @@ <h2 class="heading settled" data-level="3" id="discovery"><span class="secno">3.
...
     <dd data-md>
-     <p>The display name of the responding agent intended
-to be displayed to a user by the requesting agent.  If the responding agent
-is not yet authenticated, the requesting agent should make UI affordance for
-indicating to the user that the display name is not yet verified.  If the
-responding agent changes its display name, the requesting agent should
-make UI affordance for indicating to the user that the display name has
-changed.</p>
+     <p>The display name of the receiver, intended to be displayed to a user by the
+ controller. The controller should indicate through the UI if the receiver
```
Comment (pthatcherg) [inline index.html]: Either side can get the display name of other; it's mutual.  It's not just receiver -> controller.

File: index.bs [hunk 22]
```diff
@@ -1527,10 +1579,64 @@ agent-info-response = {
...
+  eAC3: 6
+  FLAC: 7
+  PCM: 8
+  Vorbis: 9
+)
+
+bit-rate-limits = {
+ 0: uint ; min-bit-rate
+ 1: uint ; max-bit-rate
+ 2: uint ; max-latency
```
Comment (pthatcherg) [inline index.bs]: I don't think these should be bundled together.  It's much more common to want a max than a min bitrate.

Comment (pthatcherg) [inline index.bs]: max-latency?  I don't know what that means and it seems very hard to define.  Perhaps we can address something like the sender being able to control the receiver's jitter buffer, but it doesn't seem like a capability at all.

File: index.bs (hunk 10 above)
Comment (baylesj) [reply index.bs]: %s/controller/listener/g
%s/receiver/advertiser/g

File: index.html (hunk 19 above)
Comment (baylesj) [reply index.html]: I think the PR's diff is just bunk, the log looks right to me.

File: index.bs (hunk 16 above)
Comment (baylesj) [reply index.bs]: I'll cleanup language.

File: index.bs (hunk 22 above)
Comment (baylesj) [reply index.bs]: Deleting.

File: index.html [hunk 23]
```diff
@@ -2351,10 +2790,64 @@ <h2 class="heading settled" id="appendix-a"><span class="content">Appendix A: Me
...
+  FLAC: 7
+  PCM: 8
+  Vorbis: 9
+)
+
+bit-rate-limits = {
+ 0: uint ; min-bit-rate
+ 1: uint ; max-bit-rate
+ 2: uint ; max-latency
+}
```
[the remark this answers is not in the document]
Comment (baylesj) [reply index.html]: I'll split up.

File: index.bs [hunk 24]
```diff
@@ -287,32 +287,85 @@ mechanism for metadata discovery.
...
+
+: max-audio-channels (optional)
+:: The receiver may indicate the maximum number of audio channels in abilities
+    standard audio setup, e.g. 2 for stereo, 5 or 7 for a surround setup.
+
+: audio-bit-rate-limits (optional)
+:: The receiver may indicate the practical audio bit rate limits, including the
+    minimum and maximum bit rates, and the maximum latency. Physical hardware
+    requirements, such as maximum decoder throughput or maximum audio buffer
+    size inform these values.
```
[the remark this answers is not in the document]
Comment (baylesj) [reply index.bs]: Removing audio bit rate.

File: index.bs [hunk 25]
```diff
@@ -1527,10 +1579,64 @@ agent-info-response = {
...
+
+bit-rate-limits = {
+ 0: uint ; min-bit-rate
+ 1: uint ; max-bit-rate
+ 2: uint ; max-latency
+}
+
+streaming-capabilities = {
+  0: uint ; max-pixel-fill-rate
+  1: [1* codec] ; supported-codecs
```
[the remark this answers is not in the document]
Comment (baylesj) [reply index.bs]: Changing name.

File: index.html [hunk 26]
```diff
@@ -2351,10 +2790,64 @@ <h2 class="heading settled" id="appendix-a"><span class="content">Appendix A: Me
...
+
+streaming-capabilities = {
+  0: uint ; max-pixel-fill-rate
+  1: [1* codec] ; supported-codecs
```
[the remark this answers is not in the document]
Comment (baylesj) [reply index.html]: String sounds great.

File: index.bs [hunk 27]
```diff
@@ -287,32 +287,85 @@ mechanism for metadata discovery.
...
+
+If the receiver is already authenticated, the controller has the ability to
+request additional information by sending an streaming-capabilities-request
+message, and receive back a streaming-capabilities-response message with the
+following properties:
+
+: max-pixel-fill-rate (required)
+:: The receiver shall reply with the supported pixel fill rate in megapixels per
+    second. Requesting agents can use this information to make better-educated
+    decisions about what resolution and frame rates to stream content at.
```
[the remark this answers is not in the document]
Comment (baylesj) [reply index.bs]: Added new codec type.

File: index.bs [hunk 28]
```diff
@@ -287,32 +287,85 @@ mechanism for metadata discovery.
...
+: audio-only (optional)
+:: Generally speaking, receivers are assumed to support streaming both
+    audio and video. If a receiver wishes to indicate that they only support audio
+    stream, they can pass this property as true. In that case, the controller
+    should indicate through the UI that the receiver is audio only.
+
+If the receiver is already authenticated, the controller has the ability to
+request additional information by sending an streaming-capabilities-request
+message, and receive back a streaming-capabilities-response message with the
+following properties:
```
[the remark this answers is not in the document]
Comment (baylesj) [reply index.bs]: I'll push a new revision addressing your feedback, then we can pull over.

File: index.bs [hunk 29]
```diff
@@ -287,32 +287,85 @@ mechanism for metadata discovery.
...
+:: If the receiver is a hardware device, the model name of
     the device.  This is used mainly for debugging purposes, but may be
     displayed to the user of the requesting agent.
 
-<!-- TODO: Add device type and/or capabilities -->
+: audio-only (optional)
+:: Generally speaking, receivers are assumed to support streaming both
+    audio and video. If a receiver wishes to indicate that they only support audio
+    stream, they can pass this property as true. In that case, the controller
+    should indicate through the UI that the receiver is audio only.
```
[the remark this answers is not in the document]
Comment (baylesj) [reply index.bs]: Agreed.

File: index.bs [hunk 30]
```diff
@@ -287,32 +287,85 @@ mechanism for metadata discovery.
...
+:: The receiver may indicate the practical audio bit rate limits, including the
+    minimum and maximum bit rates, and the maximum latency. Physical hardware
+    requirements, such as maximum decoder throughput or maximum audio buffer
+    size inform these values.
+
+: video-bit-rate-limits (optional)
+:: The receiver may indicate the practical audio bit rate limits, including the
+    minimum and maximum bit rates, and the maximum latency. Physical hardware
+    requirements, such as maximum decoder throughput or maximum video buffer
+    size inform these values.
```
[the remark this answers is not in the document]
Comment (baylesj) [reply index.bs]: Removed audio bit rate limits.

File: index.bs (hunk 15 above)
Comment (baylesj) [reply index.bs]: https://github.com/webscreens/openscreenprotocol/pull/151

File: index.bs [hunk 31]
```diff
@@ -297,32 +297,84 @@ mechanism for metadata discovery.
...
-unverified (such as indicating to a user that a display name of an
-unauthenticated agent is unverified).
+If a listening agent ("listener") wants to connect to an advertising agent
+("advertiser"), or to learn further metadata about it, it initiates a [[QUIC]]
+connection to the IP and port from the SRV record.  Prior to authentication,
+a message may be exchanged (such as further metadata), but such info should be
+treated as unverified (such as indicating to a user that a display name of an
+unauthenticated agent is unverified). Note that both listeners and advertisers
+may request information about each other (performing as a "requester" to
+query a "requestee").
```
Comment (mfoltzgoogle) [inline index.bs]: This paragraph is vague and confusing.  It starts by suggesting that the listener wants to learn about the advertiser.  But then says that each side can learn about the other once they are connected.   It's also unclear what message in the spec can be sent as "unverified" data.

I would suggest dropping the last sentence as it's covered in a more specific way in the next paragraph.

File: index.bs [hunk 32]
```diff
@@ -297,32 +297,84 @@ mechanism for metadata discovery.
...
 
-To learn further metadata, an agent may send an agent-info-request
-message (see [[#appendix-a]]) and receive back an agent-info-response message.  The
-messages may contain the following information with the following meaning:
+To learn further metadata, devices may send an agent-info-request
+message (see [[#appendix-a]]) and receive back an agent-info-response message.
+Note that both the listener and advertiser may send this request to learn about
```
Comment (mfoltzgoogle) [inline index.bs]: s/listener and advertiser/any agent/

File: index.bs [hunk 33]
```diff
@@ -297,32 +297,84 @@ mechanism for metadata discovery.
...
+the capabilities of the other device.
+
+The messages may contain the following properties:
```
Comment (mfoltzgoogle) [inline index.bs]: s/may// since there are required properties

Comment (mfoltzgoogle) [reply index.bs]: Are these properties of the agent-info-request or the agent-info-response?

File: index.bs (hunk 28 above)
[the remark this answers is not in the document]
Comment (mfoltzgoogle) [reply index.bs]: codecs could be used for streaming or remote playback, so I think we should keep it generic.  Unless we believe that the capabilities will be different for a given device.

[the remark this answers is not in the document]
Comment (baylesj) [reply index.bs]: Removed streaming capabilities, will fixup other patch.

File: index.bs [hunk 34]
```diff
@@ -287,32 +287,85 @@ mechanism for metadata discovery.
...
+: max-pixel-fill-rate (required)
+:: The receiver shall reply with the supported pixel fill rate in megapixels per
+    second. Requesting agents can use this information to make better-educated
+    decisions about what resolution and frame rates to stream content at.
+
+: supported-codecs (required)
+:: The receiver must indicate to the controller what codecs it understands for
+    remote playback.
+
+: min-resolution (optional)
```
[the remark this answers is not in the document]
Comment (mfoltzgoogle) [reply index.bs]: +1  I'm not sure codecs belong in a device-level metadata, but instead should be part the negotiation specifically for streaming or remote playback.  Ideally we can share some messages between the two because there's overlap in the use case: the controller has some media it wants to play on the receiver with a given codec, resolution/framerate, and bitrate.  The receiver can answer if it can play that media at all and/or play it smoothly or efficiently (see the [Media Capabilities API](https://wicg.github.io/media-capabilities/) for a Web API for querying this data).

File: index.bs [hunk 35]
```diff
@@ -287,32 +287,85 @@ mechanism for metadata discovery.
...
+: video-bit-rate-limits (optional)
+:: The receiver may indicate the practical audio bit rate limits, including the
+    minimum and maximum bit rates, and the maximum latency. Physical hardware
+    requirements, such as maximum decoder throughput or maximum video buffer
+    size inform these values.
+
+: aspect-ratio (optional)
+:: The receiver may indicate what its ideal aspect ratio is, e.g. a 16:10
+    display could return this value as 1.6 to indicate its preferred content
+    scaling.
```
[the remark this answers is not in the document]
Comment (mfoltzgoogle) [reply index.bs]: Nit, but this feels better as a rational number, because of 16:9, 4:3, etc.

File: index.bs [hunk 36]
```diff
@@ -287,32 +287,85 @@ mechanism for metadata discovery.
...
+
+: color-profiles (optional)
+:: The receiver may indicate what color profiles are understood, such as
+    sRGB v4. The controller may use these values to determine how to encode
+    video.
+
+: supports-scaling (optional)
```
[the remark this answers is not in the document]
Comment (mfoltzgoogle) [reply index.bs]: In general there are multiple native resolutions that have hardware support for scaling.  The receiver could list a set of supported resolutions and then indicate if non-standard resolutions are also supported.

File: index.bs [hunk 37]
```diff
@@ -297,32 +297,40 @@ mechanism for metadata discovery.
...
 
-<!-- TODO: Add device type and/or capabilities -->
+: supports-audio (optional)
+:: The requestee has to indicate that it supports audio streaming. If false
+    or not included, it is assumed audio streaming is not supported.
+
+: supports-video (optional)
+:: The requestee has to indicate that it supports video streaming. If false
+    or not included, it is assumed video streaming is not supported.
+
```
Comment (mfoltzgoogle) [inline index.bs]: In the future we should consider capabilities for audio input and video input as well.  But these capabilities are more often associated with removable peripherals, so it may not be suitable for device-level metadata.

Maybe we should make all capabilities playback-level metadata?  That is, part of the url-availability-request, media-request and/or remote-playback-availability-request?

File: index.bs (hunk 33 above)
Comment (baylesj) [reply index.bs]: Clarified.

File: index.bs (hunk 31 above)
Comment (baylesj) [reply index.bs]: Agreed.

File: index.bs (hunk 32 above)
Comment (baylesj) [reply index.bs]: Changed.

File: index.bs (hunk 37 above)
Comment (baylesj) [reply index.bs]: That's a good question. I'll leave this for now, I'll be in Fremont tomorrow/Friday if you want to sync up in person and hammer down some of the specifics around this.

Comment (rektide) [reply index.bs]: This is an even bigger can of worms, but I'd like to see input device support as a capability too. The trendy new big touchscreen collaborative displays come to mind as the sort of thing one would expect to openscreenprotocol into, which would have their own input device systems.

I don't know whether this capability ought be broken down into smaller pieces- `supports-input` or `supports-touch` `supports-keyboard` `supports-mouse`?

As per previous comments, maybe doesn't need to be done now? I can open a follow-up item depending on peoples feels here.

Comment (baylesj) [reply index.bs]: I would vote for a follow up item, personally. I generally prefer smaller patches when possible.

File: index.bs [hunk 38]
```diff
@@ -297,32 +297,37 @@ mechanism for metadata discovery.
...
+a display name of an unauthenticated agent is unverified).
 
 To learn further metadata, an agent may send an agent-info-request
-message (see [[#appendix-a]]) and receive back an agent-info-response message.  The
-messages may contain the following information with the following meaning:
+message (see [[#appendix-a]]) and receive back an agent-info-response message.
+Any agent may send this request to learn about the capabilities of
+another device.
+
+The agent-info-response messages contain the following properties:
```
Comment (pthatcherg) [inline index.bs]: "The X message contains" or "X messages contain", not "The X messages contain,"

File: index.bs [hunk 39]
```diff
@@ -297,32 +297,37 @@ mechanism for metadata discovery.
...
 : display-name (required)
-:: The display name of the responding agent intended
-    to be displayed to a user by the requesting agent.  If the responding agent
-    is not yet authenticated, the requesting agent should make UI affordance for
-    indicating to the user that the display name is not yet verified.  If the
-    responding agent changes its display name, the requesting agent should
-    make UI affordance for indicating to the user that the display name has
-    changed.
+:: The display name of the agent, intended to be displayed to a user by the
+     requester. The requester should indicate through the UI if the device
```
Comment (pthatcherg) [inline index.bs]: "if the device" => "if the responder"

File: index.bs [hunk 40]
```diff
@@ -297,32 +297,37 @@ mechanism for metadata discovery.
...
+     is not authenticated or if the display name changes.
 
 : model-name (optional)
 :: If the agent is a hardware device, the model name of
     the device.  This is used mainly for debugging purposes, but may be
     displayed to the user of the requesting agent.
 
-<!-- TODO: Add device type and/or capabilities -->
+: supports-audio (optional)
+:: The agent has to indicate that it supports audio streaming. If false
```
Comment (pthatcherg) [inline index.bs]: It think we should show support for audio and video in general, not specific to streaming (such as for the presentation API).  This is mostly to know if we should show an a/v icon or just an audio icon.  

Also, the name should probably be "receives-audio" and "receives-video".

Comment (baylesj) [reply index.bs]: Agreed.

File: index.bs (hunk 39 above)
Comment (baylesj) [reply index.bs]: +1

File: index.bs (hunk 38 above)
Comment (baylesj) [reply index.bs]: +1

**Comment by baylesj** [comment]
@mfoltzgoogle, thoughts on closed comments below. TL;DR I believe I have addressed all of your comments in my other patch,
https://github.com/webscreens/openscreenprotocol/pull/154

**https://github.com/webscreens/openscreenprotocol/pull/140#discussion_r272304956**
"Codecs should be generic":
_Agreed, cleaned up in second patch._

**https://github.com/webscreens/openscreenprotocol/pull/140#discussion_r272308357**
"Codecs belong in streaming":
_Agreed, moved as part of moving to second patch._

**https://github.com/webscreens/openscreenprotocol/pull/140#discussion_r272308928**
"Rational number":
_Makes sense. Added a new ratio type, since we don't appear to have one._

**https://github.com/webscreens/openscreenprotocol/pull/140#discussion_r272309518**
"Multiple native resolutions":
_Agreed, I updated to add a : native-resolutions (optional) field._

**Comment by mfoltzgoogle** [comment]
Approved the current PR with two items for follow-up discussion:

- In the `agent-info` message, using booleans versus a set-of-enums to describe capabilities
- How to describe the semantics of `receives-audio` and `receives-video` more specifically.  Does that mean an agent is required to process set of messages, a set of codecs or is it entirely up to the implementation?

Still awaiting approval from @pthatcherg.

**Comment by pthatcherg** [comment]
I think enums instead of bools is an interesting idea.  

As for "receives-audio" and "receives-video", I think the part we're missing is to have a capabilities for indicating what parts of the protocol the agent speaks.  For example, it acts as a receiver for presentation API or remote playback API or streaming or all of them.  Or it acts as a sender/controller for presentation API or remote playback API or streaming or all of them.

I think "receives-audio" and "receives-video" implies that it's a receiver, and for now that probably means both presentation API and remote playback API, but not streaming.

**Comment by pthatcherg** [comment]
I think we can land this and continue discussing in more detail.
